DEVEXP-507: add samples operator proposal for bootstrapping as removed for ipv6/d… (#172)
openshift/enhancements: enhancements/samples/removed-bootstrapping.md

updating the samples registry on the operator's config object
  -- adding any unwanted imagestreams to the skipped list of the operator's config object, or
  -- marking the operator as Removed so that no imagestreams are created and no image imports are attempted.
- for IPv6 single stack based installs, since the terms based registry, registry.redhat.io, currently does not 
support IPv6, all samples aside from the Jenkins imagestreams (which are part of the install payload), will fail to
import out of the gate, where again, unless one of the options noted for disconnected/restricted network are
employed, the operator will make itself degraded after 2 hours.

## Release Signoff Checklist

- [/] Enhancement is `implementable`
- [/] Design details are appropriately documented from clear requirements
- [/] Test plan is defined
- [/] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

In such scenarios where samples installation is doomed to fail out of the gate, it now is apparent to us that 

> the samples operator should start off a `Removed`, with no samples content installed.

Reviewer: ```suggestion
the samples operator should start off as `Removed`, with no samples content installed.
```

Author: part of next push

Those scenarios currently include
- disconnect / restricted network installs
- IPv6 single stack installs

## Motivation

> With the mandate in 4.2 that the  samples operator should mark itself `Degraded` if any imagestream imports

Reviewer: ```suggestion
With the mandate in 4.2 that the samples operator should mark itself `Degraded` if any imagestream imports
```

Author: part of next push

2) Allow the cluster administrator time to decide what to do with samples in the cluster

### Non-Goals

Other usability improvements around installing samples from registries other than registry.redhat.io have 
been identified in Jira:

- regular expressions for further manipulation of the image reference in the imagestream
- modes where failed imagestreams are automatically added to the skip list (where the administrator then
automatically can have them added to the management set by fixing the import by mirroring the images, fixing
the registry in question, etc.)
- moving the samples operator out of the payload / CVO and making it an OLM operator, where the cluster
administrator controls which imagestreams are installed when the provision the OLM operator
- aggregating network topology details like disconnected/mirrored via our cluster config API (which was previously 
discussed and ultimately abandoned).

Bootstrapping as removed is considered the meets minimum, least impacting change to the samples operator 
that addresses the scenarios where out of the box imagestreams pointing to registry.redhat.io will fail.

## Proposal

> With startup today, the samples operator bootstraps as `Managed`.  That boostrapping can be made conditional,

Reviewer: ```suggestion
With startup today, the samples operator bootstraps as `Managed`.  That bootstrapping can be made conditional,
```

Author: part of next push

> and instead come up as `Removed` when the samples operator sees

Reviewer: sees what?

Author: will change `sees` to `determines image stream imports will fail`

updated to that accessible registry in the sample operator's config object, b) Imagestreams not mirrored are added
to the config's `skippedImagestreams` list, c) the operator is switched to `Managed`
- Whatever proxy or underlying network condition that prevents access to `registry.redhat.io` is addressed and then
the samples operator is switched back to `Managed`

### Risks and Mitigations

1) Misidentifying a cluster as tolerating `Managed` out of the box (i.e. failing to catch a case we are trying
to prevent with this enhancement)
2) Bootstrapping as `Removed` for a cluster where imagestreams are importable (i.e. breaking something that works today)

Also note, a fair amount of tests in `e2e-*-builds`, `e2e-aws-image-ecosystem`, `e2e-aws-jenkins`, and the subset of 
conformance tests in e2e-aws that leverage OpenShift Builds depend on the samples as a starting point.

For e2e environments for IPv6 and disconnected / restricted network, those tests will either have to be filtered 
out of the ginkgo query or reworked to either create their own imagestreams / templates or only leverage local

> resources that do not require network access (for example, binary builds).

Reviewer: binary builds isn't really the right alternative here.  binary builds are a way to supply source code, but this is about supplying the builder images (in the absence of valid imagestreams).

Author: changing `reworked to....` to

`reworked so that any needed builder/deployment images typically retrieved from registry.redhat.io are either created or preloaded in the test cluster prior to their use in any extended test cases.`

## Design Details

### Test Plan

Consider the following in developing a test plan for this enhancement:
- QE already had some disconnected CI and we've been working with them via bugzillas to get it set up.  This change
could remove the need for additional administration on their part to keep the cluster from having `Degraded` operators
- For IPv6, the samples operator e2e will be updated to inspect the test pod for IPv6 and adjust accordingly

### Graduation Criteria

No tech preview, beta, multiple maturity levels needed.

1) Samples install as they do traditionally in a connected IPv4 environment.
2) Samples start as `Removed` in the special case scenarios noted above.

Graduation will be achieved when this support GA's minimally in 4.4 and 4.3.z as part of IPv6 initiative. 

### Upgrade / Downgrade Strategy

Samples operator already identifies migration today, along with existing vs.
new installations.

The current plan is to only target new installations with these changes.

However, an implementation where for an existing install we 

> - use our existing identifcation of whether imagestreap imports are failing

Reviewer: ```suggestion
- use our existing identification of whether imagestream imports are failing
```

Author: part of next push

- and the techniques noted above to discern whether disconnected/restricted network install or IPv6 single stack ar
in play

could allow us to take an existing installation with a samples operator either in `Degraded` state or destined to 
be in `Degraded` state and adjust from `Managed` to `Removed` automatically.

This choice is noted here for now to facilitate discussion during the enhancement review process.

### Version Skew Strategy

N/A

## Implementation History

## Drawbacks

Working from `Removed` to a working `Managed` seems the easier / better path based on initial customer engagements with
disconnected / restricted network install.  But Murphy's Law would probably say somebody has gotten used to the 
existing out of the box behavior and would rather go from a broken `Managed` to functional `Managed` operator in
one fell swoop.

> For example, as part of the multi-arch work, IBM has complained that samples exist out of the box for x86 but

Reviewer 2: @gabemontero since this is public, please drop the reference to IBM. "Early testers" or "early adopters" may be a better phrase.

Reviewer: we should also use more positive phrasing instead of "complained".

requested we add..... asked for..... etc.

we don't want to be perceived as being annoyed by our customers/users requests.

Author: reference switched @adambkaplan PTAL

Author: verb changed @bparees

Author: sorry needs another tweak just a sec

Author: OK now PTAL @bparees @adambkaplan

> they should as well for s390x or ppc64le.

Reviewer: to be fair, those adopters didn't ask for us to install content that would fail to import or otherwise fail to work, so i'm not sure how analogous the scenario is.  but yes it is indicative that there is an appetite for this content, so bootstrapping to removed may be viewed negatively by impacted users who, as you note above, may prefer to have "Broken managed content" than "no content", as their starting point.

Author: The `For example...` was added upon request from @adambkaplan in one of the earlier reviews.

Given that
- Perhaps I'm not capturing what he really wanted 
- Obviously I was originally fine with not mentioning this at all, since lack of it lead to him asking for it, and so I would not feel strongly about removing it if it is more trouble than it is worth.
- Or I could make it generic and say something along the lines of "samples are popular and not having them available and working out of the box will be viewed negatively"
- Or we can stand pat

WDYT @bparees ?

Reviewer: > Or I could make it generic and say something along the lines of "samples are popular and not having them available and working out of the box will be viewed negatively"

personally i like this one, but we can see what @adambkaplan says.

it's not going to be a sticking point for me either way.

automatically can have them added to the management set by fixing the import by mirroring the images, fixing
the registry in question, etc.)
- moving the samples operator out of the payload / CVO and making it an OLM operator, where the cluster
administrator controls which imagestreams are installed when the provision the OLM operator

Bootstrapping as removed is considered the meets minimum, least impacting change to the samples operator 
that addresses the scenarios where out of the box imagestreams pointing to registry.redhat.io will fail.

## Proposal

Out startup today, the samples operator bootstraps as `Managed`.  That boostrapping can be made conditional,
and instead come up as `Removed` when the samples operator sees

- Via the techniques noted in the "OCP 4.3 single stack IPv6 control stack" document at
https://docs.google.com/document/d/1zznvkK87y7UvOJZqs2DM4oWzx94P8Lg6YUb43whil3g/edit#heading=h.q0ahufbae2kz the 
samples operator determines that it is in an IPv6 single stack scenario
- By inspecting the image of its own pod, if the samples operator sees something other than 

> `registry.svc.ci.openshift.org` or `registry.redhat.io` assume it is a disconnected / restricted network install

Reviewer 3: Is there any acceptable scenario where there is a _non_ transparent proxy here?

Author: Only if a non transparent proxy can show up in the image reference for a pod backing a CVO operator ....  I would presume that means it shows up at the install payload level, but that is an educated guess on my part.

@ashcrow @bparees do you know if that is the case ?  Or who should we pull in to confirm one way or the other?

Author: see https://github.com/openshift/enhancements/pull/172#discussion_r366535986

scenario

> - Or (or additionally) attempt to create TCP connections to `registry.redhat.io` over the course of 1 minute and 

Reviewer 3: Thought: If the need to figure out if a cluster is restricted, disconnected, or fully connected is needed by other operators/components it may be worth having it denoted and queryable inside the cluster to avoid needing to do tests in an attempt to come to a reasonable assumption. Such an idea would be out of scope for this enhancement.

Author: We would be happy to consume such a setting if it existed in the cluster.  This approach was formulated when our own research determined one did not exist and we had to do something quickly for ipv6.

Author: turns out we've simplified things ... on bootstrap, if `registry.redhat.io` is inaccessible over a 3 minute window for any reason, we bootstrap as `Removed`, with the reason fields in the operator status indicating this situation.

The administrator can the see this and decide to leave as is or rectify via previously documented approaches.

if those fail, assume it is a disconnected / restricted network install scenario 

### User Stories [optional]

#### Story 1

#### Story 2

### Implementation Details/Notes/Constraints [optional]

### Risks and Mitigations

1) If the pod or api object inspections noted in the proposal are not definitive and misidentify some percentage of

> the scenarios 

Reviewer: there's really two risks:

1) misidentifying the situation and going to "removed" when we should not have (in particular this could break CI as well as providing a bad user experience)
2) misidentifying an going to "managed" when the imports will never succeed (this is no worse than where we are today)

we also need to consider the impact on CI if the samples operator comes up as removed (some tests may need to be reworked)

Author: yes there are 2 flavors of picking the wrong management state to bootstrap to ... did not feel the precision was necessary

that said, noting the needed CI analysis is probably worthwhile, so I'll the detailing of the risks since I'm in there

Author: section updated @bparees ptal

## Design Details

### Test Plan

Consider the following in developing a test plan for this enhancement:
- QE already had some disconnected CI and we've been working with them via bugzillas to get it set up.  This change
could remove the need for additional administration on their part to keep the cluster from having `Degraded` operators

> - For IPv6, the samples operator e2e will be updated to inspect the test pod for IPv6 and adjust accordingly

Reviewer: presumably more than just the samples operator e2e will need adjustment for CI to pass on ipv6 since no samples are going to import.  But that's true whether or not you make this change since the samples already don't import today.

Author: yep chose not to speak to broader concerns ... I'll keep this as is

  - "@adambkaplan"
  - "@derekwaynecarr"
creation-date: 2020-01-06
last-updated: 2020-01-06
status: implementable
see-also:
replaces:
superseded-by:
---

# Bootstrapping the samples operator as "Removed" on x86 for non-standard installs

Since the advent of OpenShift 4.1 and the initial introduction of the samples operator,
the range of install scenarios where attempting to install samples out of the box is counter productive
has increased with both 4.2, as well as 4.3/4.4 (which have not yet GA'ed as of this writing).

In particular:
- initial disconnected/restricted network install deployments have proven cumbersome because 
if the cluster administrator does not address the needs of Managed samples within 2 hours the operator
marks itself `Degraded`.  The current choices the cluster administrator has to avoid this are, namely:

-- mirroring images from registry.redhat.io to the mirrored registry used for install, and 
updating the samples registry on the operator's config object
-- adding any unwanted imagestreams to the skipped list of the operator's config object, or
-- marking the operator as Removed so that no imagestreams are created and no image imports are attempted.

> - for IPv6 single stack based installs, since the terms based registry, registry.redhat.io, currently does not 

Author: part of next push

support IPv6, all samples aside from the Jenkins imagestreams (which are part of the install payload), will fail to
import out of the gate, where again, unless one of the options noted for disconnected/restricted network are
employed, the operator will make itself degraded after 2 hours.

## Release Signoff Checklist

- [/] Enhancement is `implementable`
- [/] Design details are appropriately documented from clear requirements
- [/] Test plan is defined
- [/] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

> ## Open Questions [optional]

Author: removed

- since there are many imagestreams, most likely add the imagestreams they are not concerned with to the 
skip list in the operator config object
- override the registry used by the samples operator in its config object
- or if you want to punt on samples entirely, mark the operator as `Removed`

### Goals

1) Alleviate out of the install concerns in scenarios where samples installation will fail out of the gate

2) Allow the cluster administrator time to decide what to do with samples in the cluster

### Non-Goals

Other usability improvements around installing samples from registries other than registry.redhat.io have 
been identified in Jira:

- regular expressions for further manipulation of the image reference in the imagestream
- modes where failed imagestreams are automatically added to the skip list (where the administrator then
automatically can have them added to the management set by fixing the import by mirroring the images, fixing
the registry in question, etc.)
- moving the samples operator out of the payload / CVO and making it an OLM operator, where the cluster
administrator controls which imagestreams are installed when the provision the OLM operator

> 

Reviewer 2: I would also add as a non-goal: aggregating network topology details like disconnected/mirrored via our cluster config API. This is what was previously discussed and ultimately abandoned.

Author: added

Bootstrapping as removed is considered the meets minimum, least impacting change to the samples operator 
that addresses the scenarios where out of the box imagestreams pointing to registry.redhat.io will fail.

## Proposal

With startup today, the samples operator bootstraps as `Managed`.  That boostrapping can be made conditional,
and instead come up as `Removed` when the samples operator sees

- Over a 3 minute window during initial install, the samples operator will attempt to create a TCP connection 
to `registry.redhat.io` and if one never succeeds, will bootstrap as `Removed`, with no samples installed.
- The clusteroperator `Reason` fields note the inability to access the terms based registry.

The cluster administrator is then free to investigate why the inability to access `registry.redhat.com` exists

> and act accordingly.  Options could include

Reviewer 2: nit
```suggestion
and act accordingly. Options could include:
```

- Samples are left in `Removed`
- Images related to samples imagestreams are mirrored into an accessible registry, and then a) `samplesRegistry` is
updated to that accessible registry in the sample operator's config object, b) Imagestreams not mirrored are added
to the config's `skippedImagestreams` list, c) the operator is switched to `Managed`
- Whatever proxy or underlying network condition that prevents access to `registry.redhat.io` is addressed and then
the samples operator is switched back to `Managed`

> ### User Stories [optional]

Reviewer 2: Nit - remove optional sections that are not being used.

Author: removed

#### Story 1

#### Story 2

### Implementation Details/Notes/Constraints [optional]

### Risks and Mitigations

1) Misidentifying a cluster as tolerating `Managed` out of the box (i.e. failing to catch a case we are trying
to prevent with this enhancement)
2) Bootstrapping as `Removed` for a cluster where imagestreams are importable (i.e. breaking something that works today)

> Also note, a fair amount of tests in e2e-*-builds, e2e-aws-image-ecosystem, e2e-aws-jenkins, and the subset of 

Reviewer 2: nit so Markdown isn't confused:
```suggestion
Also note, a fair amount of tests in `e2e-*-builds`, `e2e-aws-image-ecosystem`, `e2e-aws-jenkins`, and the subset of 
```

conformance tests in e2e-aws that leverage OpenShift Builds depend on the samples as a starting point.

For e2e environments for IPv6 and disconnected / restricted network, those tests will either have to be filtered 
out of the ginkgo query or reworked to either create their own imagestreams / templates or only leverage local
resources that do not require network access (for example, binary builds).

## Design Details

### Test Plan

Consider the following in developing a test plan for this enhancement:
- QE already had some disconnected CI and we've been working with them via bugzillas to get it set up.  This change
could remove the need for additional administration on their part to keep the cluster from having `Degraded` operators
- For IPv6, the samples operator e2e will be updated to inspect the test pod for IPv6 and adjust accordingly

### Graduation Criteria

No tech preview, beta, multiple maturity levels needed.

1) Samples install as they do traditionally in a connected IPv4 environment.
2) Samples start as `Removed` in the special case scenarios noted above.

Graduation will be achieved when this support GA's minimally in 4.4 and 4.3.z as part of IPv6 initiative. 

> #### Examples

Reviewer 2: Nit - remove unused examples

Author: removed

### Upgrade / Downgrade Strategy

Samples operator already identifies migration today, along with existing vs.
new installations.

The current plan is to only target new installations with these changes.

However, an implementation where for an existing install we 

- use our existing identifcation of whether imagestreap imports are failing
- and the techniques noted above to discern whether disconnected/restricted network install or IPv6 single stack ar
in play

could allow us to take an existing installation with a samples operator either in `Degraded` state or destined to 
be in `Degraded` state and adjust from `Managed` to `Removed` automatically.

This choice is noted here for now to facilitate discussion during the enhancement review process.

### Version Skew Strategy

N/A

## Implementation History

## Drawbacks

Working from `Removed` to a working `Managed` seems the easier / better path based on initial customer engagements with
disconnected / restricted network install.  But Murphy's Law would probably say somebody has gotten used to the 
existing out of the box behavior and would rather go from a broken `Managed` to functional `Managed` operator in

> one fell swoop.

Reviewer 2: Another drawback (from our multi-arch conversations) - not having Samples can create the perception that customers have an incomplete installation/product.

Author: note added
